compiler/rustc_trait_selection/src/solve/mod.rs

```diff
@@ -221,6 +221,103 @@ impl<'a, 'tcx> EvalCtxt<'a, 'tcx> {
+            }
+            ty::Adt(adt, args) => {
+                // For now, unions are always considered inhabited
+                if adt.is_union() {
+                    return Err(NoSolution);
+                }
+
+                // Non-exhaustive ADTs from other crates are always considered inhabited
+                if adt.is_variant_list_non_exhaustive() && !adt.did().is_local() {
+                    return Err(NoSolution);
+                }
+
+                // The ADT is uninhabited if all its variants are uninhabited.
+                for variant in adt.variants() {
+                    if variant.is_field_list_non_exhaustive() && !variant.def_id.is_local() {
+                        // Non-exhaustive variants from other crates are always considered inhabited.
+                        return Err(NoSolution);
+                    }
+                    // This variant is uninhabited is any of its fields is uninhabited.
+                    let field_tys = variant.fields.iter().filter_map(|field| {
+                        let ty = tcx.type_of(field.did).instantiate(tcx, args);
+                        match (field.vis, module) {
+                            // The field is public or we do not care for visibility.
+                            (ty::Visibility::Public, _) | (_, None) => Ok(ty),
+                            (ty::Visibility::Restricted(from), Some(module)) => {
+                                if tcx.is_descendant_of(module, from) {
+                                    Ok(ty)
+                                } else {
+                                    // From `module`, this field may be privately uninhabited.
+                                    // Do not consider it as uninhabited.
+                                    Err(NoSolution)
+                                }
+                            }
+                        }
+                        .ok()
+                    });
+                    self.compute_uninhabited_goal_fields(param_env, module, field_tys)?;
+                }
+                self.evaluate_added_goals_and_make_canonical_response(Certainty::Yes)
+            }
+            ty::Array(ty, len) => match len.try_eval_target_usize(tcx, param_env) {
+                None | Some(0) => Err(NoSolution),
+                Some(1..) => {
+                    self.add_goal(Goal::new(
+                        tcx,
+                        param_env,
+                        ty::PredicateKind::Uninhabited(*ty, module),
+                    ));
+                    self.evaluate_added_goals_and_make_canonical_response(Certainty::Yes)
+                }
+            },
+            ty::Alias(ty::Opaque, _) => Err(NoSolution),
+            // FIXME
+            ty::Alias(ty::Projection | ty::Inherent, _) => Err(NoSolution),
+            // use a query for more complex cases
+            // references and other types are inhabited
+            _ => Err(NoSolution),
```

Comment:
Can you make this exhaustive just for good measure? We prefer exhaustive matches on `TyKind` in the new solver "just in case" :smile_cat: